lib/YaST/Firstboot/FirstbootController.pm

```diff
@@ -0,0 +1,44 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+
+# Summary: Controller for firstboot client
+# Maintainer: QE YaST <qa-sle-yast@suse.de>
+
+package YaST::Firstboot::FirstbootController;
+use strict;
+use warnings;
+use YuiRestClient;
+use YaST::Firstboot::GenericClient;
+use YaST::Firstboot::LanClient;
+use YaST::NetworkSettings::OverviewTab;
+use YaST::NetworkSettings::NetworkCardSetup::AddressTab;
+use YaST::NetworkSettings::NetworkCardSetup::GeneralTab;
+use YaST::NetworkSettings::NetworkCardSetup::VLANAddressTab;
+use testapi;
+
+sub new {
+    my ($class, $args) = @_;
+    my $self = bless {}, $class;
+    return $self->init($args);
+}
+
+sub init {
+    my ($self, $args) = @_;
+    $self->{GenericClient} = YaST::Firstboot::GenericClient->new({app => YuiRestClient::get_app()});
+    $self->{LanClient} = YaST::Firstboot::LanClient->new({app => YuiRestClient::get_app()});
+    return $self;
+}
+
+sub generic_smoketest {
+    my ($self, $debug_label) = @_;
+    $self->{GenericClient}->assert_client($debug_label);
+    $self->{GenericClient}->press_next();
```

Comment:
we are using getters to get the object instead of calling directly.

Reply:
I know, but then I'd like some explaination on this: I just don't see the point, for  me it seems to just add extra complexity and without doing anything. But I guess there's a reason I don't get ?

Reply:
those objects are private variables, so it is recommended in any programming languages to use getters, so for example we have been adding to those getter some extra functionality, is_shown() to know that something is shown in the screen, in that way we don't have to write everytime a line for that, we just have in the getter, same apply for setters in OOP. They are just principle to use, but it is true that in Perl doesn't look so obvious.

Reply:
Thanks :)

Reply:
As far as I understand, in practice, in our code, I found only one single example where it's actually useful, in [ExpertPArtionerContoller](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/bbbc31884cf8769027fef6d8b0df53487a2b7017/lib/Installation/Partitioner/LibstorageNG/v4_3/ExpertPartitionerController.pm#L96). So I wonder if we are doing something wrong, as those getters just look like  a huge amount of useless duplicated lines of codes, when they are  supposed to avoid duplicating code. As-is, it just made sense to me to get rid of it.

Reply:
Encapsulation is one of the core concepts of OOP and getters and setters and never a waste of time or a repetition, they help us to design better code. In OOP design the shortest code might not be the best (i.e.: some design pattern are very cumbersome but the server the purpose to try to avoid problems in the future). In other case, for example when you write some little algorithm inside a function, in general short amount of code is better if readable, sometimes there are idioms to use even.